File: cpp/n2_add_two_numbers.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        bool pls = false;
        ListNode *node1 = l1;
        ListNode *node2 = l2;
        ListNode *last = new ListNode();
        while(node1 != nullptr && node2 != nullptr){
            node1->val += node2->val;
            if(pls){
                node1->val += 1;
                pls = false;
            }
            if(node1->val > 9){
                node1->val -= 10;
                pls = true;
            }
            if(node1->next == nullptr){
                if(node2->next == nullptr && pls){
                    node1->next = last;
                }else{
                    node1->next = node2->next;
                    node2->next = nullptr;
                }   
            }
            node1 = node1->next;
            node2 = node2->next;
        }
        while(node1 != nullptr){
            if(pls){
                node1->val += 1;
                pls = false;
            }
            if(node1->val > 9){
                node1->val -= 10;
                pls = true;
            }
            if(node1->next == nullptr && pls){
                
                node1->next = last;
            }
            node1 = node1->next;
        }
        return l1;
    }
};

File: python/n2_add_two_numbers.py
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        node1 = l1
        node2 = l2
        pls = False
        while node1 != None and node2 != None:
            node1.val += node2.val
            if pls:
                node1.val += 1
                pls = False
            if node1.val > 9:
                node1.val -= 10
                pls = True
            if node1.next == None:
                if node2.next != None:
                    node1.next = node2.next
                    node1 = node1.next
                    print(node1.val)
                else:
                    if pls:
                        tmp = ListNode(0, None)
                        node1.next = tmp
                        node1 = node1.next
                break
            node1 = node1.next
            node2 = node2.next
        while node1 != None:
            if pls:
                node1.val += 1
                pls = False
            if node1.val > 9:
                node1.val -= 10
                pls = True
            if node1.next == None and pls:
                tmp = ListNode(0, None)
                node1.next = tmp
            node1 = node1.next
        return l1

if __name__ == "__main__":
    s = Solution()